nanoFramework.System.Net/NetworkHelper/NetworkHelper.cs

```diff
@@ -0,0 +1,300 @@
+        /// </summary>
+        private static void WaitForValidDateTime()
+        {
+            // if SNTP is available and enabled on target device this can be skipped because we should have a valid date & time
+            while (DateTime.UtcNow.Year < 2021)
+            {
+                Debug.WriteLine("Checking for valid date time...");
+
+                // wait for valid date & time
+                Thread.Sleep(1000);
+            }
+        }
+
+        /// <summary>
+        /// Checks if there is an IPAddress assigned to the device and sets the <see cref="_ipAddressAvailable"/> event accordingly.
+        /// </summary>
+        private static bool CheckIP()
+        {
+            Debug.WriteLine("Checking for IP");
+
+            var currentIPAddress = IPGlobalProperties.GetIPAddress();
+
+            if (currentIPAddress != System.Net.IPAddress.Any)
+            {
+                Debug.WriteLine($"We have an IP: {currentIPAddress}");
+                _ipAddressAvailable.Set();
+
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static void AddressChangedCallback(object sender, EventArgs e)
+        {
+            CheckIP();
+        }
+
+        /// <summary>
+        /// Perform setup of the various fields and events, along with any of the required event handlers.
+        /// </summary>
+        /// <param name="setupEvents">Set true to setup the events. Required for the thread approach. Not required for the CancelationToken implementation.</param>
+        private static void InternalSetup(bool setupEvents)
+        {
+            if (_helperInstanciated)
+            {
+                throw new InvalidOperationException();
+            }
+            else
+            {
+                // set flag
+                _helperInstanciated = true;
+
+                // setup event
+                _ipAddressAvailable = new(false);
+
+                NetworkInterface[] nis = NetworkInterface.GetAllNetworkInterfaces();
+
+                if (setupEvents)
+                {
+                    // check if there
+                    if (nis.Length == 0)
+                    {
+                        _networkHelperStatus = NetworkHelperStatus.FailedNoNetworkInterface;
+
+                        throw new NotSupportedException();
+                    }
+
+                    // setup handler
+                    NetworkChange.NetworkAddressChanged += new NetworkAddressChangedEventHandler(AddressChangedCallback);
+
+                    // instantiate events
+                    _networkReady = new(false);
+                }
+
+                foreach (var networkInterface in nis)
+                {
+                    // target only the working interface type
+                    if (networkInterface.NetworkInterfaceType == _workingNetworkInterface)
```

Comment:
you may have multiple ones and you may want to set them up differently: ex: fix IP address for ethernet and dhcp for wifi. The current logic does not allow this flexibility.

Reply:
I believe it covers the needs for Ethernet. Like I've wrote above, for Wi-Fi all those will be available there.

Reply:
If the wifi elements are in the device, it should be handy to be able to check the ip address and the rest from here. No need of additional nuget or anything else